collect/cache/cuckoo.go

```diff
@@ -0,0 +1,74 @@
+package cache
+
+import (
+	"sync"
+
+	cuckoo "github.com/panmari/cuckoofilter"
+)
+
+// This wraps a cuckoo filter implementation in a way that lets us keep it running forever without filling up.
+// You must call Maintain() periodically, most likely from a goroutine.
+// We maintain two filters that are out of sync; when current is full, future is half full and we then discard current
+// and replace it with the future, and start with a new empty future.
+
+type CuckooTraceChecker struct {
+	current   *cuckoo.Filter
+	future    *cuckoo.Filter
+	mut       sync.RWMutex
+	capacity  uint
+	firstTime bool
+}
+
+func NewCuckooTraceChecker(capacity uint) *CuckooTraceChecker {
+	return &CuckooTraceChecker{
+		capacity: capacity,
+		current:  cuckoo.NewFilter(capacity),
+		future:   cuckoo.NewFilter(capacity),
+	}
+}
+
+// Add puts a traceID into the filter.
+func (c *CuckooTraceChecker) Add(traceID string) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.current.Insert([]byte(traceID))
+	if !c.firstTime {
+		c.future.Insert([]byte(traceID))
+	}
+}
+
+// Check tests if a traceID is (very probably) in the filter.
+func (c *CuckooTraceChecker) Check(traceID string) bool {
+	b := []byte(traceID)
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+	return c.current.Lookup(b)
```

Comment:
I'm no expert in caching strategy, but when reading about cuckoo cache for this PR it sounds like cuckoo can give false positives, but never false negatives.  Do we care about false positives here?  What will be the impact of a false positive?

Reply:
This is true -- a positive says "we saw this trace before and dropped it". The effect of this means that some trace that might not have been dropped might instead be dropped. But since the decision will be consistent, it's no worse than making a negative random choice a few nanoseconds later.